presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java

```diff
@@ -110,6 +116,26 @@ public void testNonComparableWindowOrder()
         assertFails(TYPE_MISMATCH, "SELECT row_number() OVER (ORDER BY t.x) FROM (VALUES(color('red'))) AS t(x)");
     }
 
+    @Test
+    public void testORWarning()
+    {
+        assertHasWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a OR t1.b = t2.b"),
+                PERFORMANCE_WARNING, "line 1:41: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+        assertHasWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND t1.a != t2.b OR t1.b > t2.b"),
+                PERFORMANCE_WARNING, "line 1:58: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+        assertHasWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a OR IF(t2.b = t1.a OR t2.b = null, 'YES', 'NO') = 'YES'"),
+                PERFORMANCE_WARNING, "line 1:41: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+        assertHasWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a \n" + "AND (t1.b = t2.b OR t1.b > t2.b)"),
+                PERFORMANCE_WARNING, "line 2:18: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+    }
+
+    @Test void testNoORWarning()
+    {
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a"));
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND t1.b = t2.b"));
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND IF(t2.b = t1.a OR t2.b = null, 'YES', 'NO') = 'YES'"));
```

Comment:
Why should this not throw a warning? These are equivalent:
`IF(t2.b = t1.a OR t2.b = null, 'YES', 'NO') = 'YES'`
`t2.b = t1.a OR t2.b = null`

Reply:
@prithvip @prithvip it's a bit hard to do this simplification. @chi-tsai  wrote a really good test case but unfortunately we can't detect it right now and anyway it won't be a cross join because the there is still one key condition (t1.a = t2.). So I think for now it's good to err on the side of giving fewer false positives like:

T1.x = T2.x AND IF(T1.y = 'a' OR T1.z = 2, w, null) = T2.w 

I think that's ok for now.

Maybe later on, we can try and move this check to the planner where we do DNF and CNF so we can better error messages.

Reply:
@kaikalur I think that this way of warning will be confusing for users, because it is unclear what is actually being warned against. One of the test cases for the warning is this query:
`SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND (t1.b = t2.b OR t1.b > t2.b)`
There is a key condition here too, just like in your example. 
I agree that it is more difficult to do more sophisticated detection of cross joins. 
So in the absence of that, we should stick with a simpler definition for the warning, that triggers if there is any OR condition present in the join condition predicates, even if it is in a conditional or other construct.

Reply:
If we want to warn people about cross join, why don't we check this in JoinNode instead after planning? You'd potentially have some expressions pushed down to simplify it (I didn't think this through though).